### ISSUE TITLE
Implement AVOID storing what you can calculate.

### ISSUE THREAD

**Issue by JPaulsen** [dart-lang/linter#472]
From [Effective Dart](https://www.dartlang.org/guides/language/effective-dart/usage#avoid-storing-what-you-can-calculate)

**Comment by bwilkerson** [comment]
It is going to be very difficult to not have false positives with this rule. There are other considerations (such as performance) that sometimes override this advice. I'd be very interested in hearing what you're actually planning on doing to implement this one.

**Comment by JPaulsen** [comment]
I have an idea that could work, but it will require another scope (I know that is not really nice to create them and review them, sorry), what we could do is a scope that keeps dependencies of variables, so if we create something like:

```dart
constructor(int x) {
  this.x = x
  int a = this.x + 5;
}
```

Then the local variable a will depende of this.x and a literal 5, if we do this now:

```dart
this.y = a;
```

Then the field y will depende of local variable a, so recursively will depende of the field x and 5. (keeping that information in scope).

So if we solved dependencies (recursively) of a variable, and a variable only depends of literals, other field or parameters already assigned to other fields (not a non-used parameter of the constructor) then it should be linted.

But it sill being other cases, one of them is using some variables as cache, for example, we want to keep the information of an expensive function in a variable that depends of other fields we should have to replace it when we change the value of those fields, for example code likes this.

```dart
constructor(SomeType a) {
  this._x = a;
  this.y = this._x.someExpensiveMethod();
}
set x (SomeType a) {
 this._x = a;
 this.y = this._x.someExpensiveMethod();
}
```

The rule would lint this, but y is necessary because of performance, but this way to do the cache is still incorrect, the good way to do this (in my opinion) is:


```dart
constructor(SomeType a) {
  this.x = a;
}
set x (SomeType a) {
 this.x = a;
 this._y = null; // just undefine _y.
}
get y {
  if (_y == null) {
    _y = x.someExpensiveMethod();
  }
  return _y;
}
```

So the lint would be a true positive, and y should be a getter than can calculate its value.

 I think this rule would be able to help to find bugs and improve performance with the correct implementation details.

Please tell me what do you think @bwilkerson @pq

**Comment by bwilkerson** [comment]
Those are good thoughts, but a bit lower level than I was thinking of. When I look at a potential lint rule I like to try to formulate a description that could be given to a user to explain what will (and won't) be linted. For example, there is a rule that produces a lint if the `+` operator is used to concatenate two strings. It's fairly easy to explain the code pattern the rule is looking for.

I don't know of a simple explanation of what this rule would look for. You proposed

> So if ... a variable only depends of literals, other field or parameters already assigned to other fields (not a non-used parameter of the constructor) then it should be linted.

But then you gave an example that used a method invocation and suggested that we should produce a lint in that case as well.

I think it will be difficult to create a definition for this rule that will not result in false positives. The reason I say that is because the choice of whether or not to cache a computed value depends on more factors than I think we can see from a static analysis of the code. The style guide even says

> Instead, your first implementation should be ...

Which assumes that this approach isn't appropriate in all cases and that it's OK to violate this guideline sometimes.

**Comment by JPaulsen** [comment]
Yes, that method invocations makes it depend of x, is the same idea. (only depend of the field x).

**Comment by JPaulsen** [comment]
```
I think it will be difficult to create a definition for this rule that will not result in false positives. 
```
^ yes, I totally agree.
